src/Aws/Kinesis/Client/Consumer.hs

```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
 -- it.
 --
 withKinesisConsumer
-  ∷ MonadConsumer m
+  ∷ ( MonadIO m
+    , MonadBaseControl IO m
+    )
   ⇒ ConsumerKit
   → (KinesisConsumer → m α)
   → m α
-withKinesisConsumer kit inner =
-  flip runReaderT kit $ do
-    batchSize ← view ckBatchSize
-    messageQueue ← liftIO ∘ newTBQueueIO $ batchSize * 10
-
-    state ← updateStreamState CR.empty ≫= liftIO ∘ newTVarIO
-    let reshardingLoop = forever $
-          handleError (\_ → liftIO $ threadDelay 3000000) $ do
-            liftIO (readTVarIO state)
-              ≫= updateStreamState
-              ≫= liftIO ∘ atomically ∘ writeTVar state
-            liftIO $ threadDelay 10000000
-
-        producerLoop = forever $
-          handleError (\_ → liftIO $ threadDelay 2000000) $ do
-            recordsCount ← replenishMessages messageQueue state
-            liftIO ∘ threadDelay $
-              case recordsCount of
-                0 → 5000000
-                _ → 70000
-
-
-    withAsync reshardingLoop $ \reshardingHandle → do
-      link reshardingHandle
-      withAsync producerLoop $ \producerHandle → do
-        link producerHandle
-        res ← lift ∘ inner $ KinesisConsumer messageQueue state
-        return res
+withKinesisConsumer kit inner = do
+  let batchSize = kit ^. ckBatchSize
+  messageQueue ← liftIO ∘ newTBQueueIO $ batchSize * 10
+
+  state ← liftIO $ updateStreamState kit CR.empty ≫= newTVarIO
+
+  let
+    reshardingLoop =
+      forever ∘ handle (\(SomeException _) → threadDelay 3000000) $ do
```

Comment:
I am always a little concerned when I see this kind of loops, in particular in a thread. If there is chance that the system goes into a bad state, this could loop forever without doing reasonable work while the reset of the system is working under the assumption that this component does it's job. It seems that `updateStreamState` is doing network IO, which could be vulnerable to such an situation. For instance the name resolution may get into a state of permanent failure. 

If this would happen, is there a mechanism that would notify the overall system and trigger appropriate actions?